File: learnLeetCode_remote/Day17/Part1/min_cost_cut.py
from functools import lru_cache
from typing import List

# LC 1547 — Minimum Cost to Cut a Stick (Interval DP)

class Solution:
    def minCost(self, n: int, cuts: List[int]) -> int:
        cuts = sorted(cuts + [0, n])
        m = len(cuts)

        @lru_cache(None)
        def dp(l: int, r: int) -> int:
            # min cost to cut between cuts[l] and cuts[r], exclusive
            if l + 1 >= r:
                return 0
            best = float('inf')
            for i in range(l + 1, r):
                cost = cuts[r] - cuts[l] + dp(l, i) + dp(i, r)
                if cost < best:
                    best = cost
            return best

        return dp(0, m - 1)

if __name__ == "__main__":
    print(Solution().minCost(7, [1,3,4,5]))  # Expected 16

File: learnLeetCode_remote/Day17/Part1/min_cost_cut.cpp
#include <bits/stdc++.h>
using namespace std;

// LC 1547 — Minimum Cost to Cut a Stick
// Interval DP (top-down with memoization)

int minCost(int n, vector<int>& cuts){
    cuts.push_back(0); cuts.push_back(n);
    sort(cuts.begin(), cuts.end());
    int m = cuts.size();
    vector<vector<int>> memo(m, vector<int>(m, -1));
    function<int(int,int)> dp = [&](int l, int r)->int{
        if (l + 1 >= r) return 0;
        if (memo[l][r] != -1) return memo[l][r];
        int best = INT_MAX;
        for (int i = l + 1; i < r; ++i){
            int cost = cuts[r] - cuts[l] + dp(l,i) + dp(i,r);
            best = min(best, cost);
        }
        return memo[l][r] = best;
    };
    return dp(0, m-1);
}

int main(){
    vector<int> cuts = {1,3,4,5};
    cout << minCost(7, cuts) << "\n"; // 16
    return 0;
}
